sea-query-diesel/src/query.rs

```diff
@@ -0,0 +1,49 @@
+use diesel::backend::Backend;
+use diesel::connection::Connection;
+use diesel::query_builder::{AstPass, Query, QueryFragment, QueryId};
+use diesel::query_dsl::RunQueryDsl;
+use diesel::result::QueryResult;
+use diesel::sql_types::Untyped;
+
+#[must_use = "Queries are only executed when calling `load`, `get_result`, or similar."]
+pub struct SeaQuery<DB: Backend> {
+    query: String,
+    binds: Vec<Box<dyn QueryFragment<DB> + Send>>,
+}
+
+impl<DB: Backend> SeaQuery<DB> {
+    pub(crate) fn new(query: String, binds: Vec<Box<dyn QueryFragment<DB> + Send>>) -> Self {
+        SeaQuery { query, binds }
+    }
+}
+
+impl<DB> QueryFragment<DB> for SeaQuery<DB>
+where
+    DB: Backend,
+{
+    fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, DB>) -> QueryResult<()> {
+        out.unsafe_to_cache_prepared();
```

Comment:
It might be useful to only call this method depending on whether sea-query believes that a prepared statement is meaningful to cache or not. It will significantly improve performance for queries that are executed more than once.

Diesel generally tries to cache all queries that are somewhat static, which means all queries that do not contain a user input dependent component (beside specific bind parameters).

Reply:
It might be already worth to have a simple heuristics there: There are no bound paramaters -> cache the query, as it does not change for different calls.

For more targeted caching you would need to access the query ast somehow to check if there is for example no `IN` expression with a variable number of binds in there or no `INSERT` statement with a variable number of rows.

Reply:
Also there is quite a lot of detailed documentation of diesels query caching here: https://docs.diesel.rs/2.1.x/diesel/connection/statement_cache/index.html#a-primer-on-prepared-statement-caching-in-diesel

Reply:
Thanks for the info. I will keep this in mind, and later may be we can have a API that walks the AST and returns false if:

1) any custom expression used
2) any `in` and `not_in` clause
3) insert many